### PR TITLE
SLR - SL-92 part 1

### PR DESCRIPTION
**Pull request by lucatume**
Ticket: SL-92

[Related Service PR](https://github.com/the-events-calendar/event-tickets-seating-service/pull/42)

Part 1 of SL-92 work to implement the selection timer.

The implementation here is one timer per session.
The session in save in browser cookies to keep the 1:1 relationship with Tickets Commerce and WooCommerce tickets.

There can be 1+ timers+post_id couples per session (only 1 supported in TC, though).

This PR lays the base, testing and more work to come in following PRs.

Key concepts:
* each PHP view that requires the timer and logic will fire the `tec_tickets_seating_seat_selection_timer` action
* on that action the `Timer` controller will either get the token and object ID from the render context (e.g. initial render on the post with tickets) ...
* ... or it will read the **earliest exprigin** (filterable logic) token and object ID from the cookie
* PHP will render the timer on the page ...
* ... and JS will either start it on a new seat selection session (invalidating previous sessions for the token and object ID)
* ... or sync it on following pages (e.g. the checkout one)

To keep the session handling efficient and separated from main tables (e.g. options) I've created a new `Sessions` table that will store the token, object ID and reservations in an efficient manner

### CODE REVIEW

**Comment by lucatume** [comment]
Note I will be addressing testing and COOKIE notes in another PR>

File: src/Tickets/Seating/Admin/Ajax.php [hunk 1]
```diff
@@ -59,19 +79,62 @@ class Ajax extends Controller_Contract {
...
-	 * @param Seat_Types $seat_types A reference to the Seat Types service object.
+	 * @param Container    $container  A reference to the DI container object.
+	 * @param Seat_Types   $seat_types A reference to the Seat Types service object.
+	 * @param Sessions     $sessions    A reference to the Sessions table object.
+	 * @param Reservations $reservations A reference to the Reservations service object.
 	 */
-	public function __construct( Container $container, Seat_Types $seat_types ) {
+	public function __construct(
+		Container $container,
+		Seat_Types $seat_types,
```
Comment (rafsuntaskin) [inline src/Tickets/Seating/Admin/Ajax.php:112]: How about we include the [`Service`](https://github.com/the-events-calendar/event-tickets/blob/92789c48eedfb10035d57209def09d6b1cd74ff8/src/Tickets/Seating/Service/Service.php#L91) instead? In that case we will have all of `seat_type`, `layouts` and `maps` from the Service.

File: src/Tickets/Seating/Admin/Ajax.php [hunk 2]
```diff
@@ -206,4 +260,100 @@ public function invalidate_layouts_cache(): void {
...
+	 *
+	 * @since TBD
+	 *
+	 * @return void The JSON response is sent to the client.
+	 */
+	public function update_reservations() {
+		if ( ! check_ajax_referer( self::NONCE_ACTION, '_ajax_nonce', false ) ) {
+			wp_send_json_error(
+				[
+					'error' => 'Nonce verification failed',
```
Comment (rafsuntaskin) [inline src/Tickets/Seating/Admin/Ajax.php:275]:
```suggestion
					'error' => __( 'Nonce verification failed', 'event-tickets' ),
```

File: src/Tickets/Seating/Admin/Ajax.php [hunk 3]
```diff
@@ -206,4 +260,100 @@ public function invalidate_layouts_cache(): void {
...
+		$decoded = json_decode( $body, true );
+
+		if ( ! (
+			$decoded
+			&& is_array( $decoded )
+			&& isset( $decoded['token'], $decoded['reservations'] )
+		) ) {
+			wp_send_json_error(
+				[
+					'error' => 'Invalid request body',
```
Comment (rafsuntaskin) [inline src/Tickets/Seating/Admin/Ajax.php:299]:
```suggestion
					'error' => __('Invalid request body', 'event-tickets' ),
```

File: src/Tickets/Seating/Admin/Ajax.php [hunk 4]
```diff
@@ -206,4 +260,100 @@ public function invalidate_layouts_cache(): void {
...
+			);
+		}
+
+		$token        = $decoded['token'];
+		$reservations = $decoded['reservations'];
+
+		if ( ! ( $this->sessions->update_reservations( $token, $reservations ) ) ) {
+			wp_send_json_error(
+				[
+					'error' => 'Failed to update the reservations',
```
Comment (rafsuntaskin) [inline src/Tickets/Seating/Admin/Ajax.php:311]:
```suggestion
					'error' => __( 'Failed to update the reservations', 'event-tickets' ),
```

File: src/Tickets/Seating/Admin/Ajax.php [hunk 5]
```diff
@@ -206,4 +260,100 @@ public function invalidate_layouts_cache(): void {
...
+	 *
+	 * @since TBD
+	 *
+	 * @return void The JSON response is sent to the client.
+	 */
+	public function remove_reservations() {
+		if ( ! check_ajax_referer( self::NONCE_ACTION, '_ajax_nonce', false ) ) {
+			wp_send_json_error(
+				[
+					'error' => 'Nonce verification failed',
```
Comment (rafsuntaskin) [inline src/Tickets/Seating/Admin/Ajax.php:331]:
```suggestion
					'error' =>__( 'Nonce verification failed', 'event-tickets' ),
```

File: src/Tickets/Seating/Admin/Ajax.php [hunk 6]
```diff
@@ -206,4 +260,100 @@ public function invalidate_layouts_cache(): void {
...
+		$reservations_cancelled = $this->reservations->cancel(
+			$post_id,
+			$this->sessions->get_reservations_for_token( $token )
+		);
+		$token_session_deleted  = $this->sessions->delete_token_session( $token );
+
+		if ( ! ( $reservations_cancelled && $token_session_deleted ) ) {
+			wp_send_json_error(
+				[
+					'error' => 'Failed to remove the reservations',
```
Comment (rafsuntaskin) [inline src/Tickets/Seating/Admin/Ajax.php:351]:
```suggestion
					'error' => __( 'Failed to remove the reservations', 'event-tickets' ),
```

**Review by rafsuntaskin — CHANGES_REQUESTED** [review]

File: src/Tickets/Seating/Admin/Ajax.php (hunk 1 above)
Comment (lucatume) [reply src/Tickets/Seating/Admin/Ajax.php:112]: Right now the class does not require all those; just Seat Types. 
I would not pull something that is not being used yet.
If it turns out it's needed, I would change.

File: src/Tickets/Seating/Meta.php [hunk 7]
```diff
@@ -51,8 +51,8 @@ class Meta {
 	 *
 	 * @var string
 	 */
-	const META_KEY_UUID = '_tec_slr_uuid';
-	
+	public const META_KEY_UUID = '_tec_slr_uuid';
```
[the remark this answers is not in the document]
Comment (lucatume) [reply src/Tickets/Seating/Meta.php:54]: Right now I would keep the verbose version, easier to find.

**Comment by lucatume** [comment]
Superseded by #3096

**Pull request closed without merging**